### PR TITLE
Read environment variables in dump_best_config

### PR DESCRIPTION
**Pull request by himkt**
Ref. https://github.com/himkt/allennlp-optuna/issues/41

## Motivation
AllenNLP supports reading some parameters from environment variables.
Thus, a config file could contain some external variables that are not related to hyperparameters.
When invoking `dump_best_config`, we have to read environment variables as well as best-performing hyperparameters.

```jsonnet
local dropout = std.parseJson(std.extVar('dropout'));
local embedding_dim = std.parseInt(std.extVar('embedding_dim'));
local lr = std.parseJson(std.extVar('lr'));


{
  dataset_reader: {
    ...
  },
  train_data_path: std.extVar('train_data_path'),       // these are not hyperparameter but simple external parameter
  validation_data_path: std.extVar('valid_data_path'),  // which are given from environment variables
  model: {
    type: 'basic_classifier',
    text_field_embedder: {
      token_embedders: {
        tokens: {
          type: 'embedding',
          embedding_dim: embedding_dim,  // it is hyperparameter defined at the top of configuration
        },
      },
    },
    seq2vec_encoder: {
      ....
    }
  },
  data_loader: {
    batch_size: 1,
  },
  ...
}
```

## Description of the changes
- 59dbd75 reads environment variables in `dump_best_config`

### CODE REVIEW

**Comment by hvy** [comment]
Could you add unit tests or if that is difficult, is this logic verified in some other way?

**Comment by hvy** [comment]
@nzw0301 could you possibly have a look at this PR too?

**Comment by nzw0301** [comment]
Sorry, I'm afraid to let me unassign myself from reviewers :(

**Comment by hvy** [comment]
@toshihikoyanase are you able to jump in here too?

**Comment by himkt** [comment]
@hvy Sorry for the late. I added the test. PTAL :pray:

**Review by hvy — APPROVED** [review]
LGTM! A minor but I now find the name `dump_best_config` a bit misleading, though. Should we at least update the docs?

**Comment by himkt** [comment]
@hvy 
I may not understand your intention but I tried to update docstring in 108a309.
Sorry but if it doesn't meet your suggestion, please tell me in detail.

**Comment by hvy** [comment]
Looks good, thanks!

**Comment by himkt** [comment]
@toshihikoyanase Please take a look. 😸

File: tests/integration_tests/allennlp_tests/test_allennlp.py [hunk 1]
```diff
@@ -225,6 +225,39 @@ def objective(trial: optuna.Trial) -> float:
...
+        os.environ["VALID_PATH"] = "tests/integration_tests/allennlp_tests/sentences.valid"
+
+        study = optuna.create_study(direction="maximize")
+        study.optimize(objective, n_trials=1)
+
+        optuna.integration.allennlp.dump_best_config(input_config_file, output_config_file, study)
+        best_config = json.loads(_jsonnet.evaluate_file(output_config_file))
+        assert os.getenv("TRAIN_PATH") == best_config["train_data_path"]
+        assert os.getenv("VALID_PATH") == best_config["validation_data_path"]
+
```
Comment (toshihikoyanase) [inline tests/integration_tests/allennlp_tests/test_allennlp.py:259]: Could you remove the environmental variables?
They can be referred by the other test cases and we may have unexpected side-effect.

```suggestion
        assert os.getenv("TRAIN_PATH") == best_config["train_data_path"]
        assert os.getenv("VALID_PATH") == best_config["validation_data_path"]

        del os.environ["TRAIN_PATH"]
        del os.environ["VALID_PATH"]
```

**Review by toshihikoyanase — COMMENTED** [review]
I'm sorry for the delayed response. The change looks good to me, but I have a few small comments.

File: tests/integration_tests/allennlp_tests/test_allennlp.py (hunk 1 above)
Comment (himkt) [reply tests/integration_tests/allennlp_tests/test_allennlp.py:259]: I see. Thank you for the catch!

**Comment by himkt** [comment]
@toshihikoyanase Thank you for the review, I revised it based on suggestions!

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your update. LGTM!

**Pull request merged**